File: algos/sorting/insertion_sort.py
"""
In-place insertion sort
"""

def swap(L, idx1, idx2):
    temp = L[idx1]
    L[idx1] = L[idx2]
    L[idx2] = temp

def insertion_sort(L):
    for i in range(1, len(L)):
        for j in reversed(range(i)):
            # pairwise swap
            if L[j+1] < L[j]:
                swap(L, j+1, j)
            else:
                break
if __name__ == "__main__":
    L1 = [2, 4, 3, 6, 5, 1]
    insertion_sort(L1)

    expected = [1, 2, 3, 4, 5, 6]
    if L1 == expected:
        print("sorted!")
    else:
        print("not sorted!")

    print(", ".join([str(elem) for elem in L1]))

File: algos/sorting/insertionSort.ts
class Insertion {
  private static swap<T>(arr: T[], idxA: number, idxB: number) {
    const temp: T = arr[idxA]
    arr[idxA] = arr[idxB]
    arr[idxB] = temp
  }

  static sort<T>(arr: T[]): void {
    for (let i = 1; i < arr.length; i++) {
      for (let j = i - 1; j >= 0; j--) {
        if (arr[j + 1] < arr[j]) {
          this.swap(arr, j + 1, j)
        } else {
          break
        }
      }
    }
  }
}

function insertionSort<T>(arr: T[]): void {
  for (let i = 1; i < arr.length; i++) {
    for (let j = i - 1; j >= 0; j--) {
      if (arr[j + 1] < arr[j]) {
        const temp: T = arr[j + 1]
        arr[j + 1] = arr[j]
        arr[j] = temp
      } else {
        break
      }
    }
  }
}

const arrToSort1: number[] = [2, 5, 1, 3, -100, -20]
insertionSort(arrToSort1)
console.log(arrToSort1)

const arrToSort2: string[] = ["z", "b", "d", "a", "x"]
Insertion.sort(arrToSort2)
console.log(arrToSort2)
